File: typespec/common/applications.go
package common

type ApplicationState string

const (
	// Any change here should reflect in the IsValid() method too
	AppliedAppState ApplicationState = "APPLIED"

	// TODO: Remember to Reject all open applications when an Opening is closed
	RejectedAppState ApplicationState = "REJECTED"

	ShortlistedAppState ApplicationState = "SHORTLISTED"
	WithdrawnAppState   ApplicationState = "WITHDRAWN"
	ExpiredAppState     ApplicationState = "EXPIRED"
)

func (s ApplicationState) IsValid() bool {
	return s == AppliedAppState ||
		s == RejectedAppState ||
		s == ShortlistedAppState ||
		s == WithdrawnAppState ||
		s == ExpiredAppState
}

File: typespec/common/applications.ts
export type ApplicationState =
  | "APPLIED"
  | "REJECTED"
  | "SHORTLISTED"
  | "WITHDRAWN"
  | "EXPIRED";

export const ApplicationStates = {
  APPLIED: "APPLIED" as ApplicationState,
  REJECTED: "REJECTED" as ApplicationState,
  SHORTLISTED: "SHORTLISTED" as ApplicationState,
  WITHDRAWN: "WITHDRAWN" as ApplicationState,
  EXPIRED: "EXPIRED" as ApplicationState,
} as const;

export function isValidApplicationState(
  state: string
): state is ApplicationState {
  return Object.values(ApplicationStates).includes(state as ApplicationState);
}
